### PR TITLE
rgw: DNM WIP zipper loadable plugins 20220707

### PR DESCRIPTION
**Pull request by kalebskeithley**
Adds (shared) libs:
rgw_sal_rados (librgw_sal_rados) rados plugin
rgw_sal_d3n (librgw_sal_d3n) d3n plugin
rgw_sal_motr (librgw_sal_mtr) motr plugin
rgw_sal_dbstore (librgw_sal_dbstore) dbstore plugin
rados_sal library (librados_sal), below the line shlib for rgw_sal_rados and rgw_sal_d3n

future:
rgw_sal_{rados,d3n,motr,dbstore} sources move to their own respective subdirs.
rados_sal sources move to a subdir
dlopen/dlsym the plugin(s) based on config instead of hard-coded

I'm not wedded to the names so if you want different names for the new
libs I'm open to changing them.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by kalebskeithley** [comment]
Non RGWSI*:
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWBucketCtl::read_buckets_stats()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_delete_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_get_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_put_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw_init_ioctx()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWMetaSyncStatusManager::init()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw::notify::publish_reserve()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw::notify::reservation_t::~reservation_t()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWRemoteMetaLog::read_sync_status()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_Bucket'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_Log'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_REST_PSTopic_AWS'

**Comment by kalebskeithley** [comment]
RGWSI* conflicts:
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWMetadataHandlerPut_SObj::RGWMetadataHandlerPut_SObj()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_MetaBackend_Handler::Op_ManagedCtx::Op_ManagedCtx()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_MetaBackend::PutParams::~PutParams()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Role_RADOS::get_be_handler()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::get_obj()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::get_zone_svc()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::init_obj_ctx()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::read()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::ROp()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::stat()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WNOp::notify()' 
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WOp::remove()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WOp::write()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::find_zone_id_by_name()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::find_zone()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_realm() const'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_sync_policy_handler() const'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_zonegroup() const'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_zone_params() const'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_realms()'  
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_zonegroups()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_zones()'

**Comment by yuvalif** [comment]
@kalebskeithley should the lua stuff be reviewed here or in https://github.com/ceph/ceph/pull/46980 ?

File: src/rgw/rgw_sal_dbstore_plugin.cc [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                bool use_gc)
+{
+  rgw::sal::DBStore* store = new rgw::sal::DBStore();
+  DB *db = nullptr;
+
+  if (store) {
+    db = store->getDB();
+    db->set_context(cct);
+
+    if (store->initialize(cct, dpp) < 0) {
```
Comment (soumyakoduri) [inline src/rgw/rgw_sal_dbstore_plugin.cc:69]: This should set lc_thread as well - https://github.com/ceph/ceph/blob/main/src/rgw/rgw_sal.cc#L144

File: src/rgw/rgw_sal_dbstore_plugin.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      delete store;
+      return nullptr;
+    }
+  }
+
+  return store;
+}
+
+int initiateStorageProvider(void *st, const void *in_dpp, void *in_cct)
```
Comment (soumyakoduri) [inline src/rgw/rgw_sal_dbstore_plugin.cc:78]: when is this function invoked?

File: src/rgw/rgw_sal_motr_plugin.cc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  return store;
+}
+
+int initiateStorageProvider(void *st, const void *in_dpp, void *in_cct)
+{
+    rgw::sal::DBStore *store = static_cast<rgw::sal::DBStore *>(st);
+    ceph::common::CephContext *cct = static_cast<ceph::common::CephContext *>(in_cct);
+    int ret = 0;
+    DBStoreManager *dbsm = new DBStoreManager(cct);
```
Comment (soumyakoduri) [inline src/rgw/rgw_sal_motr_plugin.cc:90]: Not related to Motr. This needs cleanup.

**Pull request closed without merging**